### PR TITLE
feat(eslint-plugin): [strict-boolean-expressions] add `allowEnum` option

### LINKED ISSUE #6577: Rule proposal: No implicit enum truthiness checks
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

Enums are constrained sets of named key-value pairs. They're often used in places where developers want to be able to refer to one of those pairs by name, without caring what the value is.

Checking whether an enum-typed value is truthy, however, takes a dependency on the enum's values:

```ts
if (value) { /* ... */ }
```

I propose we add a rule that prevents truthiness checks on enum values. Similar to [`@typescript-eslint/strict-boolean-expressions`](https://typescript-eslint.io/rules/strict-boolean-expressions), it would flag any `if` statement, ternary, etc. that checks on an enum. Developers should instead use an explicit `===`/`!==` check.

### Fail Cases

```typescript
enum Theme {
    Dark,
    Light
}

function withTheme(theme) {
    if (theme) {
        console.log("🌞");
    } else {
        console.log("🌝")
    }
}
```

### Pass Cases

```typescript
enum Theme {
    Dark,
    Light
}

function withTheme(theme) {
    if (theme === Theme.Dark) {
        console.log("🌞");
    } else {
        console.log("🌝")
    }
}
```

### Additional Info

Alternately, this could be an option in [`@typescript-eslint/strict-boolean-expressions`](https://typescript-eslint.io/rules/strict-boolean-expressions). But that rule is already pretty meaty and isn't quite the same - it deals with _nullable_ values, not _implicitly falsy_ values. I think this is a distinct enough classification of issue that it deserves its own rule.

I imagine we'd want to add an `allowExplicitlyValuedEnum` option for cases where the enum is given an explicit `0` or `""` value and intentionally meant to be used in truthiness checks:

```ts
enum Theme {
  None = 0,
  Dark,
  Light
}
```

### PR DESCRIPTION
**Pull request by kozlovvski**
## PR Checklist

- [x] Addresses an existing open issue: fixes #6577
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Adds `allowEnum` option, which returns early for all enums. This rule should also report suggestions – however this is beyond the scope of this issue. I will create another issue where we can discuss what the suggestions should be. 

Additionally, this PR moves `is('nullish', 'truthy number', 'enum')` and `is('nullish', 'truthy string', 'enum')` conditions from `allowNullableEnum` scope, because it should be treated as an edge case similar to what we do with nullish truthy numbers and truthy strings when `allowString` or `allowNumber` is turned on.

### CODE REVIEW

File: packages/eslint-plugin/tests/rules/strict-boolean-expressions.test.ts [hunk 1]
```diff
@@ -133,28 +133,44 @@ ruleTester.run('strict-boolean-expressions', rule, {
...
-      options: [{ allowNullableEnum: true }],
+      options: [{ allowEnum: true }],
     },
     {
       code: `
         enum ExampleEnum {
-          This = 0,
-          That = 1,
+          This = 'one',
+          That = 'two',
```
Comment (kozlovvski) [inline packages/eslint-plugin/tests/rules/strict-boolean-expressions.test.ts:173]: This diff looks weird. I've moved the tests for `is('nullish', 'truthy number', 'enum')` and `is('nullish', 'truthy string', 'enum')` conditions to the `allowEnum` tests group and I've changed the option to be `allowEnum` instead of `allowNullableEnum`

File: packages/eslint-plugin/tests/rules/strict-boolean-expressions.test.ts [hunk 2]
```diff
@@ -164,13 +180,13 @@ ruleTester.run('strict-boolean-expressions', rule, {
...
       `,
-      options: [{ allowNullableEnum: true }],
+      options: [{ allowEnum: true }],
     },
     {
       code: `
         enum ExampleEnum {
           This = 1,
-          That = 2,
+          That = 'two',
```
Comment (kozlovvski) [inline packages/eslint-plugin/tests/rules/strict-boolean-expressions.test.ts:189]: This diff looks weird. I've moved the tests for `is('nullish', 'truthy number', 'enum')` and `is('nullish', 'truthy string', 'enum')` conditions to the `allowEnum` tests group and I've changed the option to be `allowEnum` instead of `allowNullableEnum`

File: packages/eslint-plugin/tests/rules/strict-boolean-expressions.test.ts [hunk 3]
```diff
@@ -180,13 +196,31 @@ ruleTester.run('strict-boolean-expressions', rule, {
...
+          That = 1,
+        }
+        const rand = Math.random();
+        let theEnum: ExampleEnum | null = null;
+        if (rand < 0.3) {
+          theEnum = ExampleEnum.This;
+        }
+        if (theEnum) {
+        }
+      `,
```
Comment (kozlovvski) [inline packages/eslint-plugin/tests/rules/strict-boolean-expressions.test.ts:216]: This wasn't adjusted at all

File: packages/eslint-plugin/tests/rules/strict-boolean-expressions.test.ts [hunk 4]
```diff
@@ -1227,66 +1261,6 @@ if (y) {
...
-      output: `
-        enum ExampleEnum {
-          This = 1,
-          That = 2,
-        }
-        const theEnum = Math.random() < 0.3 ? ExampleEnum.This : null;
-        if (theEnum == null) {
-        }
-      `,
-    },
```
Comment (kozlovvski) [inline packages/eslint-plugin/tests/rules/strict-boolean-expressions.test.ts:1289]: This is no longer applicable due to `is('nullish', 'truthy number', 'enum')` and `is('nullish', 'truthy string', 'enum')` conditions being treated as an known edge case for `allowEnum === true`

File: packages/eslint-plugin/src/rules/strict-boolean-expressions.ts [hunk 5]
```diff
@@ -725,13 +728,36 @@ export default util.createRule<Options, MessageId>({
...
+        is('string', 'enum') ||
+        is('truthy number', 'enum') ||
+        is('truthy string', 'enum') ||
+        is('truthy string', 'number', 'enum') ||
+        is('truthy number', 'string', 'enum') ||
+        is('truthy string', 'truthy number', 'enum') ||
+        is('string', 'number', 'enum')
       ) {
+        if (!options.allowEnum) {
+          // TODO: report errors for enums
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/strict-boolean-expressions.ts:753]: Err is this still pending?

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
@kozlovvski there's a TODO in code still? I'll wait to review till it's resolved.

Sorry for the delay in getting to this PR!

**Comment by kozlovvski** [comment]
Hey, I am closing this for now – I have trouble finding time recently and I don't want to keep a stale PR open.

**Comment by JoshuaKGoldberg** [comment]
No worries, thanks for letting us know! If you end up having time, great! If not / in the meantime, this can serve as a great reference for anybody else who's interested in tackling the issue.

For anybody else interested - please do go ahead. If you build off this code, we'll try to remember to ensure there's a [co author credit](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors) on the PR.
